test/functional/lua/vim_spec.lua

```diff
@@ -319,20 +319,99 @@ describe('lua stdlib', function()
       49,
       51,
     }
+    local indices8 = {
+      [0] = 0,
+      1,
+      2,
+      3,
+      4,
+      5,
+      6,
+      7,
+      8,
+      9,
+      10,
+      11,
+      12,
+      13,
+      14,
+      15,
+      16,
+      17,
+      18,
+      19,
+      20,
+      21,
+      22,
+      23,
+      24,
+      25,
+      26,
+      27,
+      28,
+      29,
+      30,
+      31,
+      32,
+      33,
+      34,
+      35,
+      36,
+      37,
+      38,
+      39,
+      40,
+      41,
+      42,
+      43,
+      44,
+      45,
+      46,
+      47,
+      48,
+      49,
+      50,
+      51,
+    }
     for i, k in pairs(indices32) do
       eq(k, exec_lua('return vim.str_byteindex(_G.test_text, ...)', i), i)
+      eq(k, exec_lua('return vim.str_byteindex(_G.test_text, ..., false)', i), i)
+      eq(k, exec_lua('return vim.str_byteindex(_G.test_text, ..., { encoding = "utf-32"})', i), i)
     end
     for i, k in pairs(indices16) do
       eq(k, exec_lua('return vim.str_byteindex(_G.test_text, ..., true)', i), i)
+      eq(k, exec_lua('return vim.str_byteindex(_G.test_text, ..., { encoding = "utf-16"})', i), i)
     end
-    eq(
+    for i, k in pairs(indices8) do
+      eq(k, exec_lua('return vim.str_byteindex(_G.test_text, ..., { encoding = "utf-8"})', i), i)
+    end
+    matches(
       'index out of range',
       pcall_err(exec_lua, 'return vim.str_byteindex(_G.test_text, ...)', #indices32 + 1)
     )
-    eq(
+    matches(
       'index out of range',
       pcall_err(exec_lua, 'return vim.str_byteindex(_G.test_text, ..., true)', #indices16 + 1)
     )
+    eq(
+      indices32[#indices32],
+      exec_lua(
+        'return vim.str_byteindex(_G.test_text, 99999, { encoding = "utf-32", error = false})'
+      )
+    )
+    eq(
+      indices16[#indices16],
+      exec_lua(
+        'return vim.str_byteindex(_G.test_text, 99999, { encoding = "utf-16", error = false})'
+      )
+    )
+    eq(
+      indices8[#indices8],
+      exec_lua(
+        'return vim.str_byteindex(_G.test_text, 99999, { encoding = "utf-8", error = false})'
+      )
+    )
+    eq(2, exec_lua('return vim.str_byteindex("é", 2, { encoding = "utf-16", error = false})'))
```

Comment:
With the C implementation, this will still error (even though error is false) as ```#str = 2``` which means that index goes out of range and throws from the C side